### PR TITLE
RMSNorm Blocked Implementation

### PR DESCRIPTION
**Pull request by rahulbatra85**
Switch to using blocks for columns. This gives much better performance specially for larger n_cols sizes

### CODE REVIEW

File: python/perf-kernels/rmsnorm.py [hunk 1]
```diff
@@ -46,35 +46,47 @@ def get_autotune_config():
...
-        row_start_ptr = input_ptr + row_idx * input_row_stride
+    row_idx = row_start
+
+    #Calculate squared mean by block
+    row_start_ptr = input_ptr + row_idx * input_row_stride
+    row_sum = 0.0
+    for b in tl.range(0, n_cols, BLOCK_SIZE):
+        col_offsets = b + tl.arange(0, BLOCK_SIZE)
+        input_ptrs = row_start_ptr + col_offsets
+        mask = col_offsets < n_cols
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:60]: I can't remember if we discussed it in our daily meetings, but have you tried to peel off the last loop iteration so the loop performs unmasked loads and only the last peeled iteration performs masked loads?

File: python/perf-kernels/rmsnorm.py [hunk 2]
```diff
@@ -46,35 +46,47 @@ def get_autotune_config():
...
-        row_norm = row * row  #square each value
-        row_norm = tl.sum(row_norm, axis=-1)  #sum across columns(axis=-1)
-        row_norm = row_norm / n_cols  #divide by n_cols
-        row_norm = row_norm + epsilon  #add epsilon
-        row_norm = tl.rsqrt(row_norm)  #take rsqrt, this is normalization value
-        rms_norm = row * row_norm  #multiply each x by normalization value
-        rms_norm = rms_norm * g  #element wise multiplication with g
-
-        output_row_start_ptr = output_ptr + row_idx * output_row_stride
+        mask = col_offsets < n_cols
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:74]: The peeling of last loop iteration may also be applicable here.

File: python/perf-kernels/rmsnorm.py [hunk 3]
```diff
@@ -110,6 +116,7 @@ def test_rmsnorm(M, N):
     rms_norm = torch.nn.RMSNorm(N, device='cuda')
     y_torch = rms_norm(x)
 
+    print(f"y_triton={y_triton}")
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:119]: Is it a debugging statement that must be removed before merging?

File: python/perf-kernels/rmsnorm.py [hunk 4]
```diff
@@ -87,21 +99,15 @@ def rmsnorm(x, epsilon=1e-6):
...
 
-@pytest.mark.parametrize('M, N', [
-    (1, 4),
-    (2, 10),
-    (8192, 4096),
-    (4096, 8192),
-    (1, 8192),
-    (873, 1245),
-])
+@pytest.mark.parametrize('M, N', [(1, 4), (2, 10), (8192, 4096), (4096, 8192), (1, 8192), (873, 1245), (1, 98304)])
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:110]: Was this formatting change intentional? Or it was done my `yapf`? I really think the previous formatting is much more readable. You can do something like this to disable the auto formatter:
```python
# yapf: disable
@pytest.mark.parametrize('M, N', [
    (1, 4),
    (2, 10),
    (8192, 4096),
    (4096, 8192),
    (1, 8192),
    (873, 1245),
])
# yapf: enable
```

**Review by brunomazzottiamd — APPROVED** [review]

File: python/perf-kernels/rmsnorm.py (hunk 4 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py]: This was done by yapf. Thanks for the tip!

Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py]: Fixed

File: python/perf-kernels/rmsnorm.py (hunk 1 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py]: I will try this out

File: python/perf-kernels/rmsnorm.py [hunk 5]
```diff
@@ -46,35 +46,47 @@ def get_autotune_config():
...
-def rms_kernel(output_ptr, input_ptr, g_ptr, input_row_stride, output_row_stride, n_rows, n_cols, epsilon,
+def rms_kernel(output_ptr, input_ptr, g_ptr, input_row_stride, output_row_stride, n_rows, n_cols, eps,
                BLOCK_SIZE: tl.constexpr):
     row_start = tl.program_id(0)
-    row_step = tl.num_programs(0)
-    col_offsets = tl.arange(0, BLOCK_SIZE)
-    mask = col_offsets < n_cols
-    for row_idx in tl.range(row_start, n_rows, row_step):
-        row_start_ptr = input_ptr + row_idx * input_row_stride
+    row_idx = row_start
```
[the remark this answers is not in the document]
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py]: Done

File: python/perf-kernels/rmsnorm.py (hunk 2 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py]: Done!

This helped a bit in perf!

File: python/perf-kernels/rmsnorm.py (hunk 3 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py]: Removed

**Review by brunomazzottiamd — APPROVED** [review]
